Add Quick Access extension

koinzhang opened the pull request:
## Description

- Keep your files and folders on the shelf, ready to be copied, opened or revealed in Finder.
- If the original file is deleted, the file on shelf will be deleted and not be available too.

## Screencast

https://user-images.githubusercontent.com/36128970/163184828-cbf32a94-98ab-48e1-bfbe-6272634ec4d2.mp4

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the `README` are placed outside of the `metadata` folder

thomaspaulmann commented on extensions/quick-access/package.json:
```diff
@@ -0,0 +1,125 @@
+    },
+    {
+      "name": "access-directory",
+      "title": "Access Directory",
```
I think this describes the command better.

```suggestion
      "title": "Searched Pinned Directory",
```

koinzhang replied:
> Could you add some screenshots? You can find more info [here](https://developers.raycast.com/basics/prepare-an-extension-for-store#screenshots).
 Can you swap your light/dark icons? Right now you have dark background in the dark theme. It should be light background on the dark theme to have higher contrast.
 
I forgot the metadata, I will add it.
As for the icon issue, I will try to modify it, and if it doesn't perform well after swapping icon themes, I might just keep the light-colored icons

> I think this describes the command better.

Good idea, I will modify it。

thomaspaulmann commented on extensions/quick-access/package.json:
```diff
@@ -0,0 +1,125 @@
+          "description": "Automatically copy the latest files in the topmost directory.",
+          "default": true
+        },
+        {
+          "name": "showDeleteAction",
+          "type": "checkbox",
+          "required": true,
+          "label": "Show Delete File Action",
+          "description": "Show permanent file deletion action, please check carefully.",
+          "default": true
+        },
```
Let's remove this preference. There shouldn't be any harm having this.

```suggestion
```

thomaspaulmann commented on extensions/quick-access/package.json:
```diff
@@ -0,0 +1,125 @@
+          "required": true,
+          "title": "Number of files shown",
+          "description": "Number of files displayed per directory.",
+          "default": "3",
```
Is this preference really necessary? If you keep it, I would suggest to use "All" as default option and remove the `required` option.

koinzhang replied:
Here I think it is necessary.
First let me explain the inspiration for this extension. One of the partners in the community wanted to create an extension: Copy Download Latest File, which is to copy the latest file in the download directory.

I was inspired by the idea that when we download, take a screenshot, or save a file to some common folder, like the download folder or the screenshot folder, and we want to copy or open the file immediately, we can quickly get the latest file with the help of this extension (so the files in the pinned directory are sorted by modification time).
If I only want to get the latest few files, this preference is necessary, especially in a small view window like raycast, if  show all the files, then the files in a directory may fill up the whole window, and it becomes very troublesome to find the files.

Many download files or screenshots from a long time ago are actually invalid, so it is very necessary to be able to display only the latest few files.

The functionality of this extension is similar to Raycast's newly updated Search Screnshot, except that this extension will display all types of files.

<img width="1000" alt="Quick Access" src="https://user-images.githubusercontent.com/36128970/164712178-a45331db-feac-4ef1-8cfd-88e8319ad6c4.png">
<img width="1000" alt="Search Screenshot" src="https://user-images.githubusercontent.com/36128970/164712196-570b6546-b3f0-40f1-9efb-094f7967b8d5.png">

thomaspaulmann replied:
Wouldn't sorting the files by recently modified/added solve this problem? Than the latest file is always at the top. It's just confusing that I don't see all my downloads by default. I didn't expect that and thought the extension is broken.

koinzhang replied:
Indeed it may confuse the user.
So here I will change the default to all, but the preferences will remain so that there can be more options.

thomaspaulmann replied:
Sounds good. I wouldn't make it required. Then the user can start using the extension and can optimize in the preferences later.

Generally, required preferences make mostly sense for things that need to be set up to make the extension work. This pref doesn't seem to be the case. So it's better to keep it non-required.

koinzhang replied:
In fact, I prefer to set the require to True for some preferences.
Because most of the time, users don't know that your extension has a preference option and rarely go to the preferences page to check it.

thomaspaulmann commented on extensions/quick-access/package.json:
```diff
@@ -0,0 +1,125 @@
+        {
+          "name": "sortBy",
+          "type": "dropdown",
+          "required": true,
```
Let's not make this required and pick the best default.
```suggestion
```

thomaspaulmann commented on extensions/quick-access/package.json:
```diff
@@ -0,0 +1,125 @@
+        {
+          "name": "autoCopyLatestFile",
+          "type": "checkbox",
+          "required": true,
```
Let's not make this required. Also, it's unclear what this option does.
```suggestion
```

thomaspaulmann reviewed (changes requested):
Thanks for the contribution. I left a few inline comments to make the extension easier to understand. I had a hard time to understand what it does. I think changing some naming makes it more approachable.

Few more things:
- [ ] Could you add some screenshots? You can find more info [here](https://developers.raycast.com/basics/prepare-an-extension-for-store#screenshots). 
- [ ] Can you swap your light/dark icons? Right now you have dark background in the dark theme. It should be light background on the dark theme to have higher contrast.

Thanks a lot!

thomaspaulmann commented on extensions/quick-access/README.md:
```diff
@@ -0,0 +1,28 @@
+  - Copy to Clipboard
+  - Open in Default App/Open in Finder
+  - Reveal in Finder
+  - Pin Directory
```
This action didn't work for me :-(

koinzhang replied:
Which action?
If it is "Copy File to Clipboard ", the implementation of this action uses AppleScript and accesses the file, you need to give Raycast the appropriate permissions in the system settings

thomaspaulmann replied:
The `Pin Directory`. It shows me this toast. To be honest, I also don't know what should happen. I expected that the selcted directory would be pinned and then accessible in the dropdown but it wasn't.
![CleanShot 2022-04-22 at 12 14 48](https://user-images.githubusercontent.com/12066405/164714786-95940d03-8531-4a85-a5f7-4ce50f830b49.png)

koinzhang replied:
If you want to use the Pin Directory action, you need to select a folder or, alternatively, open a Finder window first.
If you do not select any folder or open a Finder window, but use the Pin Directory action, it will detect and indicate that No directories are pinned.

koinzhang replied:
Another situation is that if you select a folder that has already been pinned, when you pin it again, since the duplicate folder will not be added, it will also prompt that No directories are pinned.

thomaspaulmann commented on extensions/quick-access/package.json:
```diff
@@ -0,0 +1,125 @@
+  "commands": [
+    {
+      "name": "pin-directory",
+      "title": "Pin Directory",
```
Is `Pin` the right name?

koinzhang replied:
The name of this extension was inspired by the Quick Access feature on windows.
So before I set the name to Quick Access and Pin Directory.
Do you have a good suggestion?

![02iOkvupvcqZ5Ybx270ZmGP-28 fit_lim size_1536x](https://user-images.githubusercontent.com/36128970/164712769-54d6d32f-bbc0-4a02-b06b-719aba415649.png)

thomaspaulmann replied:
Ah I wasn't aware of this feature. Then let's go with this name and trust Microsoft's naming ;-)

koinzhang commented:
Hi, @thomaspaulmann. Thanks for your review. Sorry to make you wait so long. I have submitted a commit to update extension.
Specific changes :

- Add metadata
- delete dark icon
- Update "Access Directoey" to "Search Pinned Directory"
- Update Preference options
- Update a bug that Toast will frequently show when you add or remove directory
- Update List Empty View

Please continue to review.

thomaspaulmann commented on extensions/quick-access/package.json:
```diff
@@ -0,0 +1,117 @@
+    },
+    {
+      "name": "search-pinned-directory",
+      "title": "Search Pinned Directory",
```
```suggestion
      "title": "Search Pinned Directories",
```

koinzhang replied:
Yes, the file name and command name have been changed to "Search Pinned Directories".

thomaspaulmann commented on extensions/quick-access/package.json:
```diff
@@ -0,0 +1,117 @@
+        {
+          "name": "autoCopyLatestFile",
+          "type": "checkbox",
+          "required": true,
```
Let's not mark this as required and pick the default that you think is best. I think not copying it automatically. Mostly because I see this screen when I open the extension for the first time and I don't know what to do here.
```suggestion
          "required": false,
```
![CleanShot 2022-04-22 at 4 04 15](https://user-images.githubusercontent.com/12066405/164741799-f44a3221-febd-4163-a4c8-6e37bdd1a63a.png)

koinzhang replied:
Okay, I'll set it to false by default and stop popping up.

thomaspaulmann commented on extensions/quick-access/src/search-pinned-directory.tsx:
```diff
@@ -0,0 +1,240 @@
+                      key={value.id}
+                      icon={isImage(parse(value.path).ext) ? { source: value.path } : { fileIcon: value.path }}
+                      title={value.name}
+                      accessories={index === 0 ? [{ icon: "🌟" }] : []}
```
This icon feels a lil out of place. If you decide to keep it, can you also add a `tooltip` to describe it?

koinzhang replied:
Maybe I can find a better way to describe it, and if I do, I'll change it; if not, I'll change 🌟 to "Latest File".

thomaspaulmann reviewed (changes requested):
Left a few minor bits. Somehow the command feels slow for what it does. I couldn't pinpoint it  but might be worth to look into it in a second iteration. Anything you think could cause it?

koinzhang commented:
Hi @thomaspaulmann . Thanks for your review.
Latest Changes:
- Update command name
- Updated preferences
- Removed "🌟"

As for the startup speed of the extension, I haven't found the reason yet, guessing that it might be the time consuming to read the folder content?

koinzhang commented:
Hi @thomaspaulmann 
I have found the cause of the slow startup and solved it, please keep reviewing.

thomaspaulmann reviewed (approved):
Great that you find the problem and fixed it. It's ready to be published. Thanks for all the effort to make this extension seamless.

PitNikola commented:
@koinzhang, thanks for building this extension. 
Quick feedback: let's find a better shortcut for deleting pinned directory than `⌘ + Delete`. This shortcut is reserved for deleting text to the beginning in the search field. I was quite confused when I wanted to delete the text and then suddenly everything was gone.

koinzhang commented:
> @koinzhang, thanks for building this extension.
> Quick feedback: let's find a better shortcut for deleting pinned directory than `⌘ + Delete`. This shortcut is reserved for deleting text to the beginning in the search field. I was quite confused when I wanted to delete the text and then suddenly everything was gone.

I didn't notice before that cmd+Backspace removes the text from the search bar, so I'm very sorry for causing this problem.
I will submit a PR to update the extension.
